### PR TITLE
Make Jaeger remote sampler available via SPI

### PR DESCRIPTION
**Pull request by pavolloffay**
Signed-off-by: Pavol Loffay <p.loffay@gmail.com>

This PR makes `JaegerRemoteSampler` available via SPI. The ultimate goal is to make it available in the java auto-instrumentation (javaagent).  I have talked about this with @tedsuo and @yurishkuro that including it in the javaagent distro should be fine (Jaeger is considered as a special vendor). 

After this PR is merged I will create a PR to auto-instrumentation to package it in the final distribution. We should also document it in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#general-sdk-configuration

### CODE REVIEW

**Comment by jkwatson** [comment]
@pavolloffay please run `./gradle build` and include changes in the docs/apidiffs in this PR, since this adds a public class to a stable module. Thanks!

**Comment by jkwatson** [comment]
@pavolloffay I'm pretty hesistant to add environment variable support for things that haven't been spec'd yet. I'm ok with adding this SPI, but without an official environment variable option, I think we shouldn't add it to the autoconfigure module yet.

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (yurishkuro) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:17]: curious why this is "jaeger service" and not the default resource.service.name

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (anuraaga) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:39]: How about moving this module to java-contrib? We're starting to use that repo more. We wouldn't need to be as strict about the spec there. Otherwise it does seem like we would want to add this Sampler to the spec.

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:17]: I was thinking about this, I will rename it.

**Comment by pavolloffay** [comment]
> @pavolloffay I'm pretty hesistant to add environment variable support for things that haven't been spec'd yet. I'm ok with adding this SPI, but without an official environment variable option, I think we shouldn't add it to the autoconfigure module yet.

Does the "without an official" mean that we can add the environment variables that are here with a note that are subject to a change or should I remove them? Note that the SPI without configuration does not make much sense. However, until the spec PR is merged I can use `OTEL_EXPORTER_JAEGER_ENDPOINT` to set the endpoint if we strictly want to use documented variables.

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:39]: The goal is to include it in the official Java agent distribution. I would say modules from `java-contrib` do not make it that far. 

Note that Jaeger remote sampler has been here in the main java repository for a long time, but it's not included in the auto-instrumentation.

File: settings.gradle.kts
Comment (anuraaga) [inline settings.gradle.kts:73]: It's already included on line 70

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (anuraaga) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:39]: Ah I didn't know that's your plan - in that case, it's even more important to go through spec. The actual sampler implementation doesn't need to be added, just add the variables

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#general-sdk-configuration

I think it could be similar to the propagators, `"jaeger": Link to Jaeger remote sampler doc` and a section with its endpoint-specific variables.

Since this artifact is already a stable one, we shouldn't add something like this name that we may need to change after the spec so can't merge this until the spec is updated

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (anuraaga) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:25]: I imagine we want this to be `initial_sampler` which can be set to any other possible value for `OTEL_SAMPLER`. Another good point we should go through the spec for

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:30]: > Ah I didn't know that's your plan - in that case, it's even more important to go through spec. The actual sampler implementation doesn't need to be added, just add the variables

yeah it is in the PR description :)

**Comment by pavolloffay** [comment]
I have updated the PR:
* added `opentelemetry-sdk-extension-jaeger-remote-sampler.txt`
* used only existing SDK configuration variables

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (anuraaga) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:30]: Yeah missed it - but while it's nice we could reuse many variables, still we've defined this new env variable value here. We need it in the spec - since you've talked to @tedsuo and @yurishkuro already presumably the spec PR can be merged relatively quickly since there already seems to be consensus for it though we'll all want to come up with a name (I wouldn't repeat the word `sampler` in this value)

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java]: removing this for now

**Comment by pavolloffay** [comment]
Spec PR https://github.com/open-telemetry/opentelemetry-specification/pull/1791

**Comment by jkwatson** [comment]
I think this will end up being fine, but let's hold off a bit until the spec gets approved, and not try to rush it into the 1.4.0 release this week.

**Review by anuraaga — APPROVED** [review]

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (pavolloffay) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:57]: can `createSampler` throw an exception? The javadoc does not prohibit it.

Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:57]: I was thinking about using `ConfigurationException` but that is not public.

**Comment by pavolloffay** [comment]
CI is failing on 

```
Run actions/setup-java@v2
Trying to resolve the latest version from remote
Error: Could not find satisfied version for SemVer '8'. 
```

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (jkwatson) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:45]: didn't this end up being "jaeger_remote" in the spec PR that got merged?

File: sdk-extensions/jaeger-remote-sampler/src/test/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProviderTest.java
Comment (anuraaga) [inline sdk-extensions/jaeger-remote-sampler/src/test/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProviderTest.java:27]:
```suggestion
    assertThat(samplerProviders).hasAtLeastOneElementOfType(JaegerRemoteSamplerProvider.class);
```

Found this while writing a similar test :)

Comment (anuraaga) [reply sdk-extensions/jaeger-remote-sampler/src/test/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProviderTest.java:27]: Or you can check the name too with this pattern

https://github.com/open-telemetry/opentelemetry-java-contrib/pull/64/files#diff-e9cb51992b77bc5666278183358a03a9c3e5b0ec6a500277e9e5681086d3f8b4R20

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (jkwatson) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:57]: exceptions thrown at this point will bubble up to the caller of auto-configuration's `initialize()` method, which I think is fine for bad configuration, since there's no way to start up an SDK in that case.

**Comment by jkwatson** [comment]
one small question, then looks good to me.

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (anuraaga) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:25]: Just noticed - can this be `config.getCommaSeparatedMap`?

Like https://github.com/open-telemetry/opentelemetry-java-contrib/pull/67/files

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:22]: good point

**Comment by pavolloffay** [comment]
PR updated

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:48]: yay I wanted to avoid getting to private fields in the test.

Comment (jkwatson) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java:48]: do you have another idea for how to test this code?

Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerProvider.java]: nope, I have added some tests using assertj. It looks rather exotic but it works :)

**Comment by pavolloffay** [comment]
I had to use `mock-maker-inline` due to final classes but then an existing test fails: 

```
> Task :sdk-extensions:jaeger-remote-sampler:test

JaegerRemoteSamplerTest > connectionWorks() FAILED
    Wanted but not invoked:
    samplingManagerImplBase.getSamplingStrategy(
        <Capturing argument>,
        <any>
    );
    -> at io.opentelemetry.sdk.extension.trace.jaeger.proto.api_v2.SamplingManagerGrpc$SamplingManagerImplBase.getSamplingStrategy(SamplingManagerGrpc.java:100)

    However, there was exactly 1 interaction with this mock:
    samplingManagerImplBase.bindService();
    -> at io.grpc.internal.ServerImplBuilder.addService(ServerImplBuilder.java:131)
        at io.opentelemetry.sdk.extension.trace.jaeger.proto.api_v2.SamplingManagerGrpc$SamplingManagerImplBase.getSamplingStrategy(SamplingManagerGrpc.java:100)
        at io.opentelemetry.sdk.extension.trace.jaeger.sampler.JaegerRemoteSamplerTest.connectionWorks(JaegerRemoteSamplerTest.java:114)


```

**Comment by yurishkuro** [comment]
🎉 🎉 🎉 

@pavolloffay once this is officially released, are you aware of any other functional gaps vs. Jaeger Java SDK? It would be nice if we could put a banner on top of Jaeger SDK repo asking people to use OTEL SDK instead.

**Pull request merged**
